packages/compiler/src/core/types.ts

```diff
@@ -2270,6 +2275,16 @@ export interface CompilerHost {
    */
   mkdirp(path: string): Promise<string | undefined>;
 
+  /**
+   * watch a file or directory for changes. 
+   * @param path path to the file/dir to watch
+   * @param onChanged callback to be called when the file/dir changes
+   */
+  watch(
```

Comment:
the file watch provided by vscode lsp is not suitable because the file we want to monitor may not in the workspace opened in vscode. (for example in the azure-rest-api-specs repo, i usually just open the folder of some service, but the package.json is actually at the root of the repo.). And I dont think it's a good idea to ask vscode to monitor all the folders from root recursively by default because recursive file watching is quite resource intense which should be used to minimal as suggested by vscode. So here i would prefer to add non-recursive watch explicitly only when needed.

And I added it into the CompilerHost because I think we are trying to control all the IO related operations there. otherwise, the functionality would not work when we replace to other virutalFs in other host. thanks.

Reply:
how do other extension do it, from what I have seen they are all trying to use the same watcher when possible. Adding it to the compiler host is bad, it brings some watching element to a single pipeline. Everything that does watch today is abstracted it different wrappper host, cli or server.

Reply:
also we do already have this case where your tsp files import files outside of the workspace do not get refreshed. As said in the other comment this is a breaking change to the host for anyone calling it programatically with a custom one. So 1. Would like to see what other extensions do here, a lot of extension have recently migrated to using the vscode watcher instead of node as its a lot more reliable(including typescript) what do they do? 2. this should be done it its own pr with review and breaking change documentation, 3. this doesn't really feel too important to watch files outside of the workspace.

Reply:
I tried the import of tsp outside of workspace which does not trigger filechanged event. I guess because we are finally using import() method to do the real work which doesn't use our filesystemcache, so it doesn't matter too much there.

I also checked the typescript. I think it only cares about the packages in the workspace scope. I tried to open a typescript project whose packages in parent folder (not in the opened folder). It just doesn't show intellisense for them anymore. But I think we can't do that because service teams are one of our major customers and I know many of them would just open the service foldder in the azure-rest-api-specs repo. I believe we would want this feature to be available to them.

On the other side. I would agree that this doesn't worth a breaking change, so what I would try to do is:
1. When starting lsp server, instead of adding a watch on the root folder, I will iterate all the parent folder of workspace folders and add a non-recursive watch for package.json in each folder. I think the number of watchers created in this way should be limited
2. Because the packages under node_modules may not be monitored (for those in workspace, they may be excluded because node_modules are in file.excludeWatch settings which is honored by vscode watch, like mine...; for those outside of workspace, we are not monitor them anyway), so 
     - in npm-package-provider, when it's notified that a pakcage.json is changed, it will try to reset itself as well as those packages in its node_modules. 
     - I may also add some handling in emitter provider to further reduce the potential IO increased. 
     - For the FileSystemCache, it may not be notified about any changes in the node_modules folder as mentioned, but I guess it should be fine considering we are using import() too, need to double check on that.

so in this way, the pros. is we are using vscode watch everywhere and no breaking change. the cons. is there would be  more IO when user update their npm packages. Overall, i think it should be fine. Will give it a try next week.

A few more thoughts about the CompilerHost and how it's used in ServerHost, I found we are using CompilerHost not just for compile but also other places (i.e. when needs to access IO) which works now but feels not what it is designed for. Feels like to me it makes more sense to provide these functionalities from the FileService which then to be used by CompilerHost and other places when needed? In this way, we can handle the case when some other functionality is needed but doesn't make sense to add to CompilerHost which is for compile. Of course I don't mean to refactor it in this PR,  but may be something we can think about in the future when needed. thanks.

Reply:
I tried the import of tsp outside of workspace which does not trigger filechanged event. I guess because we are finally using import() method to do the real work which doesn't use our filesystemcache, so it doesn't matter too much there.

I also checked the typescript. I think it only cares about the packages in the workspace scope. I tried to open a typescript project whose packages in parent folder (not in the opened folder). It just doesn't show intellisense for them anymore. But I think we can't do that because service teams are one of our major customers and I know many of them would just open the service foldder in the azure-rest-api-specs repo. I believe we would want this feature to be available to them.

On the other side. I would agree that this doesn't worth a breaking change, so what I would try to do is:
1. When starting lsp server, instead of adding a watch on the root folder, I will iterate all the parent folder of workspace folders and add a non-recursive watch for package.json in each folder. I think the number of watchers created in this way should be limited
2. Because the packages under node_modules may not be monitored (for those in workspace, they may be excluded because node_modules are in file.excludeWatch settings which is honored by vscode watch, like mine...; for those outside of workspace, we are not monitor them anyway), so 
     - in npm-package-provider, when it's notified that a pakcage.json is changed, it will try to reset itself as well as those packages in its node_modules. 
     - I may also add some handling in emitter provider to further reduce the potential IO increased. 
     - For the FileSystemCache, it may not be notified about any changes in the node_modules folder as mentioned, but I guess it should be fine considering we are using import() too, need to double check on that.

so in this way, the pros. is we are using vscode watch everywhere and no breaking change. the cons. is there would be  more IO when user update their npm packages. Overall, i think it should be fine. Will give it a try next week.

A few more thoughts about the CompilerHost and how it's used in ServerHost, I found we are using CompilerHost not just for compile but also other places (i.e. when needs to access IO) which works now but feels not what it is designed for. Feels like to me it makes more sense to provide these functionalities from the FileService which then to be used by CompilerHost and other places when needed? In this way, we can handle the case when some other functionality is needed but doesn't make sense to add to CompilerHost which is for compile. Of course I don't mean to refactor it in this PR,  but may be something we can think about in the future when needed. thanks.

Reply:
Yeah the import of tsp outside does use the file system cache(only the import of js files is delegated to node and that one won't reload ever - separate issue).

For item `1.` do you think its really worth it to even watch what is happening above? I also just tried with typescript and it won't reload if you import files outside of the scope and don't have them open in the current workspace. (Try adding `import {Bar } from "../bar.mjs"`) in a `main.js` at the root of a vscode window. And in the parent dir add `bar.mjs` and toggle adding `export class Bar {}`. THe main.js will not update to know if `Bar` is exported.

For 2. I believe what works today and shouldn't be dropped is that if you edit the tsp in a node_modules it will update(JS will not same as above). So you can use that to sometimes debug a library inline.

For the compiler host part, I am not fully sure I understand, are you saying we are calling the compiler host/server host when we should call the higher abstraction in file service? If so then yeah maybe. If you are saying we shouldn't reuse the compiler host at all in the server because its only meant for the compilation then its not really true. Its the data that abstract any os/fs specific api that are required by the compiler. Server has its own extension of it and same does the playground in the browser(not just different implementation but extra methods too).
If you think you can refactor to make some things cleaner feel free to make those PRs!

Reply:
PR updated. Please help to check again. I would prefer to monitor above because I think our azure services teams would be that user case when working on azure-rest-api-specs repo, which is one of our major customers and actually the cost is not high. (actually it's more straight forward than updating the compilerHost .....). 

Yeah, i mean we should call the higher abstraction in file service instead of compiler host when it's not for compiling. Will consider refactor that part later when necessary. thanks.